lisa/tools/ethtool.py

```diff
@@ -0,0 +1,199 @@
+            )
+
+        current_param = self.__channel_count_param_pattern.match(
+            current_settings.group("settings")
+        )
+        max_param = self.__channel_count_param_pattern.match(
+            max_settings.group("settings")
+        )
+        if (not current_param) or (not max_param):
+            raise LisaException("Cannot get device channel current and/or max count")
+
+        self.device_name = interface
+        self.current_channels = int(current_param.group("value"))
+        self.max_channels = int(max_param.group("value"))
+
+
+class DeviceFeatures:
+    # ethtool device feature info is in format -
+    # ~$ ethtool -k eth0
+    #       Features for eth0:
+    #       rx-checksumming: on
+    #       tx-checksumming: on
+    #           tx-checksum-ipv4: on
+    #           tx-checksum-ip-generic: off [fixed]
+    #           tx-checksum-ipv6: on
+    #           tx-checksum-fcoe-crc: off [fixed]
+    #           tx-checksum-sctp: off [fixed]
+    #         scatter-gather: on
+    __feature_info_pattern = re.compile(
+        r"Features for (?P<interface>[\w]*):[\s]*(?P<value>.*?)$", re.DOTALL
+    )
+    __feature_settings_pattern = re.compile(
+        r"^[\s]*(?P<name>.*):(?P<value>.*?)?$", re.MULTILINE
+    )
+
+    def __init__(self, device: str, device_feature_raw: str) -> None:
+        self._parse_feature_info(device, device_feature_raw)
+
+    def _parse_feature_info(self, device: str, raw_str: str) -> None:
+        matched_features_info = self.__feature_info_pattern.match(raw_str)
+        if not matched_features_info:
+            raise LisaException("Cannot get device's features settings info")
+
+        self.device_name = device
+        self.enabled_features = []
+        for row in matched_features_info.group("value").splitlines():
+            feature_info = self.__feature_settings_pattern.match(row)
+            if not feature_info:
+                raise LisaException(
+                    "Could not get feature setting in the defined pattern."
+                )
+            if "on" in feature_info.group("value"):
+                self.enabled_features.append(feature_info.group("name"))
+
+
+class Ethtool(Tool):
+    @property
+    def command(self) -> str:
+        return "ethtool"
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        self._command = "ethtool"
+        self._device_list: List[str] = []
+        self._device_channel_map: Dict[str, DeviceChannel] = {}
+        self._device_features_map: Dict[str, DeviceFeatures] = {}
+
+    def _install(self) -> bool:
+        posix_os: Posix = cast(Posix, self.node.os)
+        posix_os.install_packages("ethtool")
+        return self._check_exists()
+
+    def _get_device_list(self) -> None:
+        if self._device_list:
+            return
+
+        result = self.node.execute(
```

Comment:
the `self.node.shell` support file system operations, try to use those APIs.

Reply:
We may need a `find` tool.

Reply:
> We may need a `find` tool.

I am just thinking here, find is used with so many options (like here also), so we basically just pass all the options from here as a string?

Reply:
We don't provide them by options, or strings. The implementation on distros may be different. We implement them by what's needed, for example, find a pattern in a folder, `find_files(self, pattern: str, recursive: bool = True)`

Reply:
> We don't provide them by options, or strings. The implementation on distros may be different. We implement them by what's needed, for example, find a pattern in a folder, `find_files(self, pattern: str, recursive: bool = True)`

Then how will it be useful here for this case where I want to find the pattern only if it exists under another pattern. I will have to make multiple find_files() call and then I want to find in such a way that file or directory names containing spaces or newlines are  handled properly (here with option -print0, it makes sure filenames with spaces and newline characters are accomodated by adding null character at the end of the filenames)

Reply:
An equative command is like "find /sys/devices -ipath '*vmbus*/*/net' -print0". The "-print0" is used to prevent '\n' in name. If it's needed, it can be added inside of the "find" tool. If multiple condition is really necessary, we can implement another method called `find_files_by_conditions`.

Add one more argument to handle case sensitive.
`find_files(self, pattern: str, ignore_case: bool = False)`

Reply:
yeah basically would need to add all these options in this find tool. I am fine with that, just that people might keep on adding the options depending on what they would need.
So I just want to understand if we are getting any perf benefit or something via going this route against the node.execute("") way. Or it is more for like clarity?

Reply:
The tools are used for sharing the logic. There may be different behavior cross distros, even freebsd, the tool can help share the logic about compatibility. And like the logic here, it needs to split strings by 0x00, and check if there is result, those logic can be shared too.

Reply:
I am not sure what you mean by compatibility. Tools like find have same implementation across distros I believe. They offer same options to be used 
https://www.freebsd.org/cgi/man.cgi?find(1)
https://man7.org/linux/man-pages/man1/find.1.html

But I get the part about reusability, the thing is find command offer so many options and the combination of those options, not sure how much helpful it would be.
Anyways I will add the find tool.

Reply:
We don't implement all options at first time. we just implement what's used in our test code, and enrich it incrementally.

Reply:
I'm also on the bandwagon of finding tools abstraction a little of a nuisance, when it comes to very rich in options tools people are trying to encapsulate here. For find, I'd basically expose the bare executable and have the person add what they want in the run() call, given the myriad of ways it can be exercised--just a shortcut for the lazy ones not wanting to node.execute(find) themselves.